src/qibolab/identifier.py

```diff
@@ -20,44 +18,29 @@ class ChannelType(str, Enum):
     PROBE = "probe"
     ACQUISITION = "acquisition"
     DRIVE = "drive"
-    DRIVE12 = "drive12"
-    DRIVE_CROSS = "drive_cross"
     FLUX = "flux"
 
     def __str__(self) -> str:
         return str(self.value)
 
 
-class ChannelId(Model):
-    """Unique identifier for a channel."""
-
-    qubit: QubitId
-    channel_type: ChannelType
-    cross: Optional[str]
-
-    @model_validator(mode="before")
-    @classmethod
-    def _load(cls, ch: str) -> dict:
-        elements = ch.split("/")
-        # TODO: replace with pattern matching, once py3.9 will be abandoned
-        if len(elements) > 3:
-            raise ValueError()
-        q = TypeAdapter(QubitId).validate_python(elements[0])
-        ct = ChannelType(elements[1])
-        assert len(elements) == 2 or ct is ChannelType.DRIVE_CROSS
-        dc = elements[2] if len(elements) == 3 else None
-        return dict(qubit=q, channel_type=ct, cross=dc)
-
-    @classmethod
-    def load(cls, value: str):
-        """Unpack from string."""
-        return cls.model_validate(value)
-
-    def __str__(self):
-        """Represent as its joint components."""
-        return "/".join(str(el[1]) for el in self if el[1] is not None)
-
-    @model_serializer
-    def _dump(self) -> str:
-        """Prepare for serialization."""
-        return str(self)
+ChannelId = str
+"""Unique identifier for a channel."""
+
+
+StateId = int
+"""State identifier."""
+
+TransitionId = Annotated[
+    tuple[StateId, StateId],
+    BeforeValidator(lambda p: tuple(p.split("-")) if isinstance(p, str) else p),
+    PlainSerializer(lambda p: f"{p[0]}-{p[1]}"),
+]
+"""Identifier for a state transition."""
+
+QubitPairId = Annotated[
+    tuple[QubitId, QubitId],
+    BeforeValidator(lambda p: tuple(p.split("-")) if isinstance(p, str) else p),
+    PlainSerializer(lambda p: f"{p[0]}-{p[1]}"),
+]
+"""Two-qubit active interaction identifier."""
```

Comment:
Yes, I'm always in doubt when things are small enough, but copying is always hurting me a bit.

I did it because I first thought about making a function that was returning the full thing, just given a single input (that would have been `StateId` or `QubitId`). But this may not be supported by static analyzers, that are not truly executing Python code (they only support a subset of Python features).

However, while the type is used at analysis time, the annotations are only needed at runtime (and aliases would be supported also by the analyzers, even if it's not needed).
So, instead of giving up completely, I can certainly reduce the copy to the minimum, with something like:
```suggestion
x = BeforeValidator(lambda p: tuple(p.split("-")) if isinstance(p, str) else p)
y = PlainSerializer(lambda p: f"{p[0]}-{p[1]}")
TransitionId = Annotated[tuple[StateId, StateId], x, y]
"""Identifier for a state transition."""

QubitPairId = Annotated[tuple[QubitId, QubitId], x, y]
"""Two-qubit active interaction identifier."""
```

Thanks for giving me a chance to have a second look at it ;)

Reply:
Done in cf250ffc0ea79e6fc63ad9bf7bdc2dd052742fc7